cilium/CFP-33716-packet-tracing.md

```diff
@@ -0,0 +1,636 @@
+
+IPSec/Wireguard - Status: in investigation
+We expect packet tracing to be compatible with IPSec and Wireguard. These protocols encrypt packets between known 
+nodes in a cluster, but each node also has its own encryption key-pair to encrypt/decrypt traffic to and from. As
+the ip_trace_id is a series of bytes, in the packet, we expect this to be compatible with IPSec/Wireguard.
+
+## Future Milestones
+
+### Deferred Milestone 1: Use conntrack to restore trace ID for reply packets
+- When a packet includes a trace ID, the conntrack entry should record the ID along with the connection information.
+- For subsequent packets – if a trace ID is not present in the IP Options – use the track ID from conntrack to restore the internal state.
+- Note this will not modify the IP option header.
+
+### Deferred Milestone 2: Tag packets
+- **Audience:** Proficient Cilium user with knowledge of cilium-dbg / pwru tools.
+- Provide a `cilium-dbg` subcommand for a user to define what traffic should be traced.
+- Prepend tag program to container lxc interface.
+- **Optional:** Provide a flag to optionally remove the tag before egressing Node NIC, addressing targets that do not support IP options.
+- **Challenge:** This affects the available MTU.
+  - [Preferred] Only apply to packets that have the length available (e.g., packets with no data; SYN w/o fast open, RST, etc.).
+  - Modify the MTU/route MTU of the container (i.e., subtracting the IP option + trace ID lengths).
+
+### Deferred Milestone 3: Tag packets via CR
+- **Audience:** Network specialist with knowledge of network topology.
+- Develop a K8s API where the user specifies traffic to tag using common selectors: 4-tuple, pod identity, etc.
+
+## Appendix
+
+### Parsing attempted approaches
+
+#### [Nonviable] Option 2: Parse at XDP hook
+**Pros:**
+- Earliest detection of trace IDs.
+- Immediate handling of packets with trace IDs.
+
+**Considerations:**
+- XDP programs don’t support the SKB struct.
+- Must ensure that TC hook is able to parse from the xdp_md struct created in the XDP hook, leading to increased complexity.
+
+### Trace ID storage attempted approaches
+
+#### [Nonviable] Option 2: Use sk_buff extensions
+**Pros:**
+- Can extend the sk_buff structure without modifying the core structure.
+- Data freed when skb is freed.
+
+**Considerations:**
+- Compatible with kernel versions post-v5.0-rc1.
+
+#### [Nonviable] Option 3.1: Use skb->mark or skb->CB field
+**Pros:**
+- The “mark” field is a generic packet mark for sk_buff and suitable for such purposes as adding a trace ID.
+
+**Considerations:**
+- Potential conflicts with other parts of the network that use this field.
+
+#### [Nonviable] Option 3.2: Store reference into skb->mark or skb->CB field
```

Comment:
What's the difference between Option 3.1 vs 3.2?

Reply:
This was a consideration very early in investigations about the resource usage of storing the value itself, or a reference to that value, so a memory address. Given that using the mark field wasn't viable, neither could be really considered. For purposes of documentation, I removed the 3.2 option as it doesn't add much value to this CFP.